lib/shuttle/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use tokio_util::sync::CancellationToken;
+
+mod app_state;
+mod handlers;
+mod middleware;
+
+/// The header key used to indicate to a running shuttle instance that it has consumed everything
+/// and can shut down. The header value and message body are ignored.
+pub const FINAL_MESSAGE_HEADER_KEY: &str = "X-Final-Message";
+
+#[allow(missing_docs)]
+#[remain::sorted]
+#[derive(Debug, Error)]
+pub enum ShuttleError {
+    #[error("async nats consumer error: {0}")]
+    AsyncNatsConsumer(#[from] async_nats::error::Error<ConsumerErrorKind>),
+    #[error("async nats stream error: {0}")]
+    AsyncNatsStream(#[from] async_nats::error::Error<StreamErrorKind>),
+    #[error("naxum error: {0}")]
+    Naxum(#[source] io::Error),
+}
+
+type Result<T> = std::result::Result<T, ShuttleError>;
+
+/// A running, opinionated [`naxum`] server that "shuttles" messages from a limits-based stream to
+/// another given subject.
+pub struct Shuttle {
+    source_subject: Subject,
+    destination_subject: Subject,
+    inner: Box<dyn Future<Output = io::Result<()>> + Unpin + Send>,
+}
+
+impl std::fmt::Debug for Shuttle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Shuttle")
+            .field("source_subject", &self.source_subject)
+            .field("destination_subject", &self.destination_subject)
+            .finish_non_exhaustive()
+    }
+}
+
+impl Shuttle {
+    /// Creates a new running [`Shuttle`] instance.
+    pub async fn new(
+        nats: NatsClient,
+        token: CancellationToken,
+        limits_based_source_stream: async_nats::jetstream::stream::Stream,
+        source_subject: Subject,
+        destination_subject: Subject,
+    ) -> Result<Self> {
+        let deliver_subject = nats.new_inbox();
+        let connection_metadata = nats.metadata_clone();
+        let context = jetstream::new(nats);
+
+        let incoming = {
+            limits_based_source_stream
+                .create_consumer(async_nats::jetstream::consumer::push::OrderedConfig {
+                    name: Some(format!("shuttle-{}", Ulid::new())),
+                    deliver_subject,
+                    filter_subject: source_subject.to_string(),
+                    ..Default::default()
+                })
+                .await?
+                .messages()
+                .await?
+        };
+
+        let state =
+            crate::app_state::AppState::new(context, destination_subject.clone(), token.clone());
+
+        let app = ServiceBuilder::new()
+            .layer(
+                TraceLayer::new()
+                    .make_span_with(
+                        telemetry_nats::NatsMakeSpan::builder(connection_metadata).build(),
+                    )
+                    .on_response(telemetry_nats::NatsOnResponse::new()),
+            )
+            .layer(
+                PostProcessLayer::new()
+                    .on_success(DeleteMessageOnSuccess::new(limits_based_source_stream)),
```

Comment:
Stolen from `rebaser-server` for limits-based stream draining with push ordered consumers